File: homework01/caesar.py
import string

ord_islower_ind = {True: 97, False: 65}
alphabet_length = 26


def encrypt_caesar(plaintext: str, shift: int = 3) -> str:
    """
    >>> encrypt_caesar("PYTHON")
    'SBWKRQ'
    >>> encrypt_caesar("python")
    'sbwkrq'
    >>> encrypt_caesar("Python3.6")
    'Sbwkrq3.6'
    >>> encrypt_caesar("")
    ''
    """
    ciphertext = ''
    for letter in plaintext:
        if letter.isalpha():
            ciphertext += chr(
                (ord(letter) - ord_islower_ind[letter.islower()] + shift) % alphabet_length + ord_islower_ind[
                    letter.islower()])
        else:
            ciphertext += letter
    return ciphertext


def decrypt_caesar(ciphertext: str, shift: int = 3) -> str:
    """
    >>> decrypt_caesar("SBWKRQ")
    'PYTHON'
    >>> decrypt_caesar("sbwkrq")
    'python'
    >>> decrypt_caesar("Sbwkrq3.6")
    'Python3.6'
    >>> decrypt_caesar("")
    ''
    """
    plaintext = ''
    for letter in ciphertext:
        if letter.isalpha():
            plaintext += chr(
                (ord(letter) - ord_islower_ind[letter.islower()] - shift) % alphabet_length + ord_islower_ind[
                    letter.islower()])
        else:
            plaintext += letter
    return plaintext


if __name__ == '__main__':
    original = string.ascii_letters
    encrypted = encrypt_caesar(string.ascii_letters)
    decrypted = decrypt_caesar(encrypt_caesar(string.ascii_letters))

    print(original)
    print(encrypted)
    print(decrypted)
    print(f'Decrypted is equal to orig: {original == decrypted}')

File: homework01-go/caesar/caesar.go
package caesar

func EncryptCaesar(plaintext string, shift int) string {
	var ciphertext string

	for i := 0; i < len(plaintext); i++ {
		letter := plaintext[i]
		if (int('a') <= int(letter)) && (int(letter) <= int('z')) {
			ciphertext += string(rune(int(letter)-97+shift)%26 + 97)
		} else if (int('A') <= int(letter)) && (int(letter) <= int('Z')) {
			ciphertext += string(rune(int(letter)-65+shift)%26 + 65)
		} else {
			ciphertext += string(letter)
		}
	}

	return ciphertext
}

func DecryptCaesar(ciphertext string, shift int) string {
	var plaintext string

	for i := 0; i < len(ciphertext); i++ {
		letter := ciphertext[i]
		if (int('a') <= int(letter)) && (int(letter) <= int('z')) {
			newLetter := int(letter) - shift
			if newLetter < 97 {
				newLetter += 26
			}
			plaintext += string(rune(newLetter))
		} else if (int('A') <= int(letter)) && (int(letter) <= int('Z')) {
			newLetter := int(letter) - shift
			if newLetter < 65 {
				newLetter += 26
			}
			plaintext += string(rune(newLetter))
		} else {
			plaintext += string(letter)
		}
	}

	return plaintext
}
